format/macho/macho.go

```diff
@@ -0,0 +1,492 @@
+	LC_UUID:                     "LC_UUID",
+	LC_RPATH:                    "LC_RPATH",
+	LC_CODE_SIGNATURE:           "LC_CODE_SIGNATURE",
+	LC_SEGMENT_SPLIT_INFO:       "LC_SEGMENT_SPLIT_INFO",
+	LC_REEXPORT_DYLIB:           "LC_REEXPORT_DYLIB",
+	LC_LAZY_LOAD_DYLIB:          "LC_LAZY_LOAD_DYLIB",
+	LC_ENCRYPTION_INFO:          "LC_ENCRYPTION_INFO",
+	LC_DYLD_INFO:                "LC_DYLD_INFO",
+	LC_DYLD_INFO_ONLY:           "LC_DYLD_INFO_ONLY",
+	LC_LOAD_UPWARD_DYLIB:        "LC_LOAD_UPWARD_DYLIB",
+	LC_VERSION_MIN_MACOSX:       "LC_VERSION_MIN_MACOSX",
+	LC_VERSION_MIN_IPHONEOS:     "LC_VERSION_MIN_IPHONEOS",
+	LC_FUNCTION_STARTS:          "LC_FUNCTION_STARTS",
+	LC_DYLD_ENVIRONMENT:         "LC_DYLD_ENVIRONMENT",
+	LC_MAIN:                     "LC_MAIN",
+	LC_DATA_IN_CODE:             "LC_DATA_IN_CODE",
+	LC_SOURCE_VERSION:           "LC_SOURCE_VERSION",
+	LC_DYLIB_CODE_SIGN_DRS:      "LC_DYLIB_CODE_SIGN_DRS",
+	LC_ENCRYPTION_INFO_64:       "LC_ENCRYPTION_INFO_64",
+	LC_LINKER_OPTION:            "LC_LINKER_OPTION",
+	LC_LINKER_OPTIMIZATION_HINT: "LC_LINKER_OPTIMIZATION_HINT",
+	LC_VERSION_MIN_TVOS:         "LC_VERSION_MIN_TVOS",
+	LC_VERSION_MIN_WATCHOS:      "LC_VERSION_MIN_WATCHOS",
+	LC_NOTE:                     "LC_NOTE",
+	LC_BUILD_VERSION:            "LC_BUILD_VERSION",
+}
+
+func machoDecode(d *decode.D, in interface{}) interface{} {
+	var archBits int
+	var ncmds uint64
+	magicBuffer := d.U32LE()
+
+	if magicBuffer == MH_MAGIC || magicBuffer == MH_MAGIC_64 {
+		d.Endian = decode.LittleEndian
+		if magicBuffer == MH_MAGIC {
+			archBits = 32
+		} else {
+			archBits = 64
+		}
+	} else if magicBuffer == MH_CIGAM || magicBuffer == MH_CIGAM_64 {
+		d.Endian = decode.BigEndian
+		if magicBuffer == MH_CIGAM {
+			archBits = 32
+		} else {
+			archBits = 64
+		}
+	} else {
+		d.Fatalf("Invalid magic field")
+	}
+
+	d.SeekAbs(0)
+	d.FieldStruct(fmt.Sprintf("mach_header_%d", archBits), func(d *decode.D) {
+		d.FieldU32("magic", magicToHex, classBits, endianNames)
+		d.FieldS32("cputype", cpuTypes)
+		d.FieldS32("cpusubtype")
+		// TODO ask about how to symmap this as it depends on a pair of values
+		d.FieldU32("filetype") // TODO expand this
+		ncmds = d.FieldU32("ncdms")
+		d.FieldU32("sizeofncdms")
+		d.FieldU32("flags") // TODO expand flags
+		if archBits == 64 {
+			d.FieldRawLen("reserved", 4*8, d.BitBufIsZero())
+		}
+	})
+	ncmdsIdx := 0
+	d.FieldStructArrayLoop("load_command", "load_command", func() bool {
+		return ncmdsIdx < int(ncmds)
+	}, func(d *decode.D) {
+		cmd := d.FieldU32("cmd", loadCommands)
+		cmdsize := d.FieldU32("cmdsize")
+		switch cmd {
+		case LC_UUID:
+			d.FieldStruct("uuid_command", func(d *decode.D) {
+				d.FieldRawLen("uuid", 16*8)
+			})
+		case LC_SEGMENT, LC_SEGMENT_64:
+			// nsect := (cmdsize - uint64(archBits)) / uint64(archBits)
+			var nsects uint64
+			if archBits == 32 {
+				d.FieldStruct("segment_command", func(d *decode.D) {
+					d.FieldRawLen("segname", 16*8)
```

Comment:
Not a string?

Reply:
This is embarrassing but at the start I could not get the hang of it. Sad to see this being trickled down. Fixed

Reply:
👍  no need to feel embarrassed :)